src/capability.ml

```diff
@@ -0,0 +1,216 @@
+    let inh = String.contains letters 'i' in
+    let indicated_capability =
+      Capability.create ~effective:eff ~permitted:per ~inherited:inh
+    in
+    compute_capability indicated_capability all_caps operator)
+;;
+
+let split_indication_from_cap_string clause =
+  let operators = [ '+'; '-'; '=' ] in
+  let indication =
+    String.lstrip clause ~drop:(fun c -> not (List.mem operators c ~equal:Char.equal))
+  in
+  let cap_string = String.chop_suffix_exn clause ~suffix:indication in
+  indication, cap_string
+;;
+
+type t =
+  { all_caps : Capability.t
+  ; caps_by_tag : Capability.t String.Map.t
+  }
+
+let process_capabilities input_text : t =
+  let (all_caps : Capability.t), assignment_sets =
+    split_init_rest_capabilities input_text
+  in
+  let caps_by_tag =
+    List.concat_map assignment_sets ~f:(fun assign_set ->
+      let cap_clauses = String.split assign_set ~on:',' |> List.rev in
+      let indication, indicator_cap_string =
+        let indicator_clause = List.hd_exn cap_clauses in
+        split_indication_from_cap_string indicator_clause
+      in
+      let cap_strings = List.tl_exn cap_clauses in
+      let final_capability : Capability.t = get_final_cap all_caps indication in
+      List.map (indicator_cap_string :: cap_strings) ~f:(fun cap_string ->
+        cap_string, final_capability))
+    |> String.Map.of_alist_exn
+  in
+  { all_caps; caps_by_tag }
+;;
+
+let contains_cap t cap_string =
+  List.mem (Map.keys t.caps_by_tag) cap_string ~equal:String.equal
+;;
+
+let get_cap_sets t cap_string =
+  if contains_cap t cap_string
+  then Map.find t.caps_by_tag cap_string |> Option.value ~default:t.all_caps
+  else Capability.empty
+;;
+
+let has_eff t cap_string = get_cap_sets t cap_string |> Capability.effective
+
+let linux_version_has_cap_perfmon linux_version =
+  let split_version = String.split linux_version ~on:'.' in
+  let head = Int.of_string (List.hd_exn split_version) in
+  (Int.equal head 5 && Int.of_string (List.nth_exn split_version 1) >= 8) || head > 5
+;;
+
+let does_not_remove_effective t cap_string =
+  if contains_cap t cap_string then has_eff t cap_string else true
+;;
+
+let check_perf_support getcap linux_version =
+  let stripped_getcap = String.strip getcap in
+  if String.equal stripped_getcap ""
+  then false
+  else if not (String.contains stripped_getcap ' ')
+  then
+    failwith
+      "Invalid capability string" (* must not have capability if contains no spaces *)
+  else (
+    let caps = process_capabilities getcap in
+    if Capability.effective caps.all_caps (* if all_caps has eff *)
+    then
+      if linux_version_has_cap_perfmon linux_version
+      then
+        if does_not_remove_effective caps "cap_perfmon"
+        then true
+        else does_not_remove_effective caps "cap_sys_admin"
+      else does_not_remove_effective caps "cap_sys_admin"
+    else has_eff caps "cap_perfmon" || has_eff caps "cap_sys_admin")
+;;
+
+let run_test test_string linux_version =
```

Comment:
Let's move this into the test module, and name at least `~linux_version` (otherwise it's too easy to accidentally pass `test_string` in place of `linux_version`).